### PR TITLE
Add properties to QgsHighlight for easier animations

### PR DESCRIPTION
**Pull request by elpaso**

### CODE REVIEW

**Review by nyalldawson — APPROVED** [review]

**Comment by m-kuhn** [comment]
Can I ask where you are heading to with this?

File: src/gui/qgshighlight.h [hunk 1]
```diff
@@ -46,8 +46,15 @@ class QgsSymbol;
  *   highlight.show()
  * \endcode
  */
-class GUI_EXPORT QgsHighlight: public QgsMapCanvasItem
+class GUI_EXPORT QgsHighlight: public QObject, public QgsMapCanvasItem
 {
+
+    Q_OBJECT
```
Comment (m-kuhn) [inline src/gui/qgshighlight.h]: Use `Q_GADGET` if there are no signals and slots.

**Comment by elpaso** [comment]
@m-kuhn I'm going to use this from a python plugin, the Qt animation framework works well with the Qt properties.

**Comment by m-kuhn** [comment]
I see, so API only, no app changes planned?
It's pure curiosity, I like those property changes!

File: src/gui/qgshighlight.h [hunk 2]
```diff
@@ -125,6 +156,9 @@ class GUI_EXPORT QgsHighlight: public QgsMapCanvasItem
     void paintLine( QPainter *p, QgsPolylineXY line );
     void paintPolygon( QPainter *p, const QgsPolygonXY &polygon );
 
+    int mWidth = 1; // line / stroke width property
+    QColor mColor; // line / stroke color property
+    QColor mFillColor; // line / stroke fillColor property
```
Comment (m-kuhn) [inline src/gui/qgshighlight.h]: Are there no getters for these on pen and brush? Or are these performance considerations?

**Comment by elpaso** [comment]
@m-kuhn API only for now, but I'm pretty sure that someone will eventually make good use of this with nice effects.

File: src/gui/qgshighlight.h (hunk 2 above)
Comment (elpaso) [reply src/gui/qgshighlight.h]: What do you mean: forward the property to the brush?

Comment (m-kuhn) [reply src/gui/qgshighlight.h]: No, use `return mBrush->width();` in the getter.

Comment (elpaso) [reply src/gui/qgshighlight.h]: ok, got it, yes it should be possible to use the getters but maybe faster this way.

Comment (m-kuhn) [reply src/gui/qgshighlight.h]: Right, on the other hand I guess the setter is called more often (and time critical) and there it's additional overhead.
But it's just nitpicking really, I'm leaving the decision to you.

**Review by m-kuhn — CHANGES_REQUESTED** [review]
Good job. Just switch to Q_GADGET and this is good to be merged.

**Comment by elpaso** [comment]
@m-kuhn I switched to Q_GADGET but it started to crash in sip (apaprently q_metacall when trying to write the property, but my debugger does not help much here), do I also need to remove the QObject inheritance? According to http://doc.qt.io/qt-5/properties.html the class MUST inherit from QObject and according to https://doc.qt.io/archives/qt-5.8/qobject.html#Q_GADGET the macro is for classes that do NOT inherit from QObject. 

Do I miss something here?

**Comment by m-kuhn** [comment]
Yes, remove QObject

**Comment by elpaso** [comment]
@m-kuhn if I remove QObject then I will not be able to use the `QPropertyAnimation` which takes a QObject* as first argument, I think I need to stick to QObject inheritance and Q_OBJECT macro.

**Comment by elpaso** [comment]
See: http://doc.qt.io/qt-5/qpropertyanimation.html#QPropertyAnimation-1

**Comment by m-kuhn** [comment]
I guess they were too lazy to add the lightweight version of QPropertyAnimation yet.
In this case, ok, go for the QObject based approach. Does the animation work without changed signals on the property?

**Review by m-kuhn — APPROVED** [review]

**Comment by elpaso** [comment]
@m-kuhn yes: the animation works fine without changed signals declared in Q_PROPERTY

**Comment by m-kuhn** [comment]
I think it would be possible to copy the QPropertyAnimation code and make it Q_GADGET compatible. Or maybe even subclass it.

Basically it would need a void* pointer to a Q_GADGET and the name of the property to get it's index and to be call it on the QMetaProperty object.
On the other hand, that's probably not code we should be implementing in QGIS but rather code that should be integrated upstream.

https://git.merproject.org/mer-core/qtbase/commit/d9e8571f4de47d3587de07aaff71eadff771f1e2#45bc32b28f01bf9dce0edaa32fc60950a5134e61_2972_2983

**Pull request merged**